### PR TITLE
Implemented the first eigenvalue example

### PR DESCRIPTION
**Pull request by fdkong**
I have created the following hierarchy:

```
             +--> DisplacedSystem
             | 
SystemBase ->+--> AuxiliarySystem
             |
             +-->NonlinearSystemBase+--> NonlinearSystem
                                    |
                                    +--> NonlinearEigenSystem
```

```
SubProblem ->+--> DisplacedProblem
             |
             +-->FEProblemBase+--> FEProblem
                              |
                              +--> EigenProblem
```

And integrated with SLEPc to create the first standard eigenvalue example.

Before I further add more functionalities, and I would ask suggestions and comments from @idaholab/moose-team.  The implementation will affect the MOOSE a lot.  I want to know if it is a `MOOSE` way to write the code like this. And any improvements on this implementation?

### CODE REVIEW

**Comment by YaqiWang** [comment]
@fdkong Just to make sure in you first 11 commit, you are doing exactly as the message says but not sneaking anything in. 

How about we change the `NonlinearSystemBase` back to `NonlinearSystem` and rename `NonlinearSystem` to `NonlinearEquationSystem`?

I will look into it in more detail.

**Comment by YaqiWang** [comment]
I can suggest to get the names by the member functions on the level. Just for example, if one level is constructing the mesh, we can name it `MeshProblem`.

**Comment by fdkong** [comment]
> I can suggest to get the names by the member functions on the level. Just for example, if one level is constructing the mesh, we can name it MeshProblem.

According suggestions from @YaqiWang  and @jwpeterson, I am going to change `EquationProblem` to `FEProblem` (keep the same API), and `FEProblem` to `MeshProblem`. `SubProblem` will be merged into `Problem`.

@idaholab/moose-team, everybody agrees on this? @permcody, @friedmud, @andrsd ?

**Comment by friedmud** [comment]
Please don't change the interface for the class called `FEProblem`.  A LOT of user code relies on it (including LOTS of my own).

To change something like that would require a lengthy deprecation period (6 months?)... with a clear migration path.

Lots of people depend on MOOSE now... many outside the apps you guys see daily.  Gone is the time when we could make major API changes and just "deal with" the fall out.

**Comment by fdkong** [comment]
> Please don't change the interface for the class called FEProblem. A LOT of user code relies on it (including LOTS of my own).
> 
> To change something like that would require a lengthy deprecation period (6 months?)... with a clear migration path.
> 
> Lots of people depend on MOOSE now... many outside the apps you guys see daily. Gone is the time when we could make major API changes and just "deal with" the fall out.

OK, @jwpeterson has pointed out yesterday, and I will keep the same API. Do you have a suggestion on what name we have for the basis class of `FEProblem` and `EigenProblem`?

**Comment by permcody** [comment]
I haven't really been following this thread but I know this is a lot of work and I think it's looking pretty good. Thanks for responding to everyone's comments.

File: framework/src/base/EigenProblem.C
Comment (andrsd) [inline framework/src/base/EigenProblem.C]: 1 space after `=`.

File: framework/src/base/EigenProblem.C
Comment (andrsd) [inline framework/src/base/EigenProblem.C]: Both `_nl` and `_aux` could be set in the initializer list...

File: framework/src/base/EigenProblem.C
Comment (andrsd) [inline framework/src/base/EigenProblem.C]: I believe that `LIBMESH_HAVE_SLEPC` implies `LIBMESH_HAVE_PETSC`

File: framework/src/base/EigenProblem.C
Comment (andrsd) [inline framework/src/base/EigenProblem.C]: You can merge this into one `if`

File: framework/src/base/EigenProblem.C
Comment (andrsd) [inline framework/src/base/EigenProblem.C]: Spaces around `<<`

File: framework/src/base/EquationProblem.C
Comment (andrsd) [inline framework/src/base/EquationProblem.C]: Spaces around `?`

File: framework/src/base/EquationProblem.C
Comment (andrsd) [inline framework/src/base/EquationProblem.C]: Both `_nl` and `_aux` can be set in the initializer list.

File: framework/src/base/FEProblem.C
Comment (andrsd) [inline framework/src/base/FEProblem.C]: Space after a `,`

File: framework/src/base/NonlinearSystem.C
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: Indetation

File: framework/src/outputs/VariableResidualNormsDebugOutput.C
Comment (andrsd) [inline framework/src/outputs/VariableResidualNormsDebugOutput.C]: Space after `,`s

File: framework/src/preconditioners/PhysicsBasedPreconditioner.C
Comment (andrsd) [inline framework/src/preconditioners/PhysicsBasedPreconditioner.C]: Space after `,`s. Also below

File: framework/src/timesteppers/DT2.C
Comment (andrsd) [inline framework/src/timesteppers/DT2.C]: `&nl_sys`, no need for additional space, similarly below

File: framework/src/timesteppers/DT2.C
Comment (andrsd) [inline framework/src/timesteppers/DT2.C]: `*nl_sys`

File: framework/src/utils/PetscSupport.C
Comment (andrsd) [inline framework/src/utils/PetscSupport.C]: `{` on the new line

File: test/tests/problems/eigen_problem/ipm.i
Comment (andrsd) [inline test/tests/problems/eigen_problem/ipm.i]: Indentation

File: test/tests/problems/eigen_problem/ipm.i
Comment (andrsd) [inline test/tests/problems/eigen_problem/ipm.i]: Remove this

File: test/tests/problems/eigen_problem/ipm.i
Comment (andrsd) [inline test/tests/problems/eigen_problem/ipm.i]: Remove

File: test/tests/problems/eigen_problem/ipm.i
Comment (andrsd) [inline test/tests/problems/eigen_problem/ipm.i]: Remove

File: test/tests/problems/eigen_problem/ipm.i
Comment (andrsd) [inline test/tests/problems/eigen_problem/ipm.i]: `;`s?

**Review by andrsd — CHANGES_REQUESTED** [review]
I went roughly over the code, not in gore details.
1. I am not a big fan of the private members in the new FEProblem-derived classes.
2. I did not comment on the class name change that was already discussed, because I agree with the outcome.

**Comment by andrsd** [comment]
Also, I saw quite a bit of `not implemented yet`, what is the plan there?

**Comment by fdkong** [comment]
> Also, I saw quite a bit of not implemented yet, what is the plan there?

There are somethings I highly possibly (not certainly ) will  touch, and they are actually pure virtual functions in the basis class, and I don't want them keep silent at this point if they are called in other places.

File: framework/include/base/EigenProblem.h
Comment (YaqiWang) [inline framework/include/base/EigenProblem.h]: Our current eigen solver does not depend on SLEPc. We have a decision to make here, should we keep this SLEPc independent solver on our side or push it to SLEPc and make SLEPc mandatory for solving eigenvalue problems?

**Comment by fdkong** [comment]
> When people do not have SLEPc build, we should provide an "empty" EigenProblem class which will throw an error in a c-tor saying something like "You are trying to solve an eigenvalue problem, but you do not have SLEPc built in. Go recompile libmesh with such and such options."

Good suggestion, and I will do this.

**Comment by fdkong** [comment]
> Our current eigen solver does not depend on SLEPc. We have a decision to make here, should we keep this SLEPc independent solver on our side or push it to SLEPc and make SLEPc mandatory for solving eigenvalue problems?

@YaqiWang, if you want to keep the current eigen-solver on our side, I am fine with that, but it is difficult for me to maintain the current implementation.   I prefer to rewrite the current algorithms (inverse power iteration and Newton) based on SLPEc.

**Comment by YaqiWang** [comment]
Understood. This will make SLPEc required by Rattlesnake. I am not sure if it is a good idea for political reasons. Just want to warn you. We need others' opinion especially @permcody @friedmud. I personally do not like this but am willing to be persuaded.

**Comment by fdkong** [comment]
> Understood. This will make SLPEc required by Rattlesnake. I am not sure if it is a good idea for political reasons. Just want to warn you. We need others' opinion especially @permcody @friedmud. I personally do not like this but am willing to be persuaded.

If for some reasons we are not willing to include SLEPc as part of MOOSE-based applications, I definitely can build some eigenvalue solvers  based on PETSc only (of course, more work for me).  Current SLEPc also has some issues on the nonlinear eigenvalue problems, that is, it does not support eigen-vector dependent nonlinear eigenvalue  problems. More precisely, operators can not depend on the eigen vector. I have to create new algorithms, even though  we bring in SLEPc. Of course, benefit here is we could use some components from SLEPc when implementing the new algorithms. Another benefit is that we have a lot more algorithms than right now if we bring in SLEPc.

Ping @friedmud, @permcody.

**Comment by friedmud** [comment]
I don't see a reason to lose our current eigenvalue solver capability.
It's very mature and can solve very difficult problems.

Power iterations are key in many nuclear reactor applications... we're
always going to need the ability to do that manually.  For instance, I use
a power iteration based eigenvalue solver in my new method of
characteristics code.  Also: acceleration methods like CMFD require power
iterations.

Not everything we do can boil down to calling slepc.  I think we want to
_add_ slepc as a new option to see how it performs with some problems.

**Comment by fdkong** [comment]
> I don't see a reason to lose our current eigenvalue solver capability. It's very mature and can solve very difficult problems. 

The reason is that it is not easy to maintain the current implementation. I am not going to remove the current capability until SLEPc option can do the same thing. 

If @YaqiWang  does NOT like make his apps depend on SLEPc, I could re-implement the two algorithms (inverse and Newton) based on PETSc only. 

If @friedmud  still want to keep the current implementation after I have re-implemented the same algorithms using PETSc or SLEPc, I am fine too because I don't consider they can't exist together.  

>Power iterations are key in many nuclear reactor applications... we're always going to need the ability to do that manually. For instance, I use a power iteration based eigenvalue solver in my new method of characteristics code. Also: acceleration methods like CMFD require power iterations. Not everything we do can boil down to calling slepc. I think we want to _add_ slepc as a new option to see how it performs with some problems.

**Comment by friedmud** [comment]
If you can really do all of the things that the current EigenSystem does
then it should be fine.

Note: I don't actually use the current EigenSystem stuff (and I agree that
it's complicated)... I just don't want to lose sight of the fact that it
solves things that I don't think slepc can do on its own... and I would
hate to see that capability go away.

**Comment by fdkong** [comment]
> If you can really do all of the things that the current EigenSystem does then it should be fine. Note: I don't actually use the current EigenSystem stuff (and I agree that it's complicated)... I just don't want to lose sight of the fact that it solves things that I don't think slepc can do on its own... and I would hate to see that capability go away.

I know SLEPc can not do that, but I am going to build something on top of SLEPc and PETSc to make the capability. The capability we have right now will NOT go away, and will be enhanced, otherwise no reason to do this development.

**Comment by fdkong** [comment]
> Both _nl and _aux could be set in the initializer list...

There are some data members from basis class, so we could not put them in the initializer list, (I am not sure).

**Comment by andrsd** [comment]
> There are some data members from basis class, so we could not put them in the initializer list, (I am not sure).

Yeah, I saw that after I commented on it. One way how to deal with it is to move the data members into child classes and make the parent class use some virtual method to get the right instance of nl or aux... Just an idea...

**Comment by fdkong** [comment]
@permcody, @andrsd, finally passed all tests.

**Comment by permcody** [comment]
@fdkong - we'll review this. However, we should also go ahead and run "clang-format" on the files affected by this PR before we merge (as your last commit). You may as well keep the ownership of these lines since we'll run that command on the whole repo very soon. @rwcarlsen can probably help with that.

**Comment by fdkong** [comment]
> @fdkong - we'll review this. However, we should also go ahead and run "clang-format" on the files affected by this PR before we merge (as your last commit). You may as well keep the ownership of these lines since we'll run that command on the whole repo very soon. @rwcarlsen can probably help with that.

We delay to this because clang-format can create annoying things

**Comment by fdkong** [comment]
@friedmud, more suggestions on this?

File: framework/include/base/EigenProblem.h
Comment (andrsd) [inline framework/include/base/EigenProblem.h]: space after `*`

File: framework/include/splits/Split.h
Comment (andrsd) [inline framework/include/splits/Split.h]: Space before `&`.

File: framework/src/actions/ActionWarehouse.C
Comment (andrsd) [inline framework/src/actions/ActionWarehouse.C]: No space inside `<...>`

File: framework/src/actions/AddOutputAction.C
Comment (andrsd) [inline framework/src/actions/AddOutputAction.C]: 2 spaces in front of `_problem.get()`

File: framework/src/base/MooseApp.C
Comment (andrsd) [inline framework/src/base/MooseApp.C]: Spaces around `=`

File: framework/src/base/NonlinearEigenSystem.C
Comment (andrsd) [inline framework/src/base/NonlinearEigenSystem.C]: 4 spaces for the initializer list...

File: framework/src/base/NonlinearEigenSystem.C
Comment (andrsd) [inline framework/src/base/NonlinearEigenSystem.C]: Incorrect indentation. << 2

File: framework/src/base/NonlinearEigenSystem.C
Comment (andrsd) [inline framework/src/base/NonlinearEigenSystem.C]: Indent by 1 space?

File: framework/src/multiapps/MultiApp.C
Comment (andrsd) [inline framework/src/multiapps/MultiApp.C]: `FeProblem &`

File: framework/src/outputs/VariableResidualNormsDebugOutput.C
Comment (andrsd) [inline framework/src/outputs/VariableResidualNormsDebugOutput.C]: The other comma needs a space after it as well...

File: framework/src/restart/Restartable.C
Comment (andrsd) [inline framework/src/restart/Restartable.C]: This needs better indentation to increase legibility. I propose:
```
_var(bool1 ? long_thing :
  (bool2 ? longer_thing :
    (bool3 ? the_longer_things : what ever)))
```
But anything else that will clearly show what is going on with the nesting will work too. Also, since it is a pointer a nested `if` in the `c-tor` could look a lot better.

File: framework/src/timesteppers/TimeStepper.C
Comment (andrsd) [inline framework/src/timesteppers/TimeStepper.C]: I assuming this is here for backwards compatibility?

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by andrsd** [comment]
One comment on the `EigenSystem` and using preprocessor. The following typically worked better for me:

In .h:
```
#if HAVE_A
  declare class X as if we have A
#else
  declare class X as if we do not have A 
    - keep to c-tor signature the same, but does not require inheritance
    - does not even need to have the same interface!
#endif
```

In .C
```
#if HAVE_A
  implement class X as if we have A
#else
  implement class X as if we do not have A
#endif
```

This obviously depends on how we want X to behave if we don't have A. If our goal is to just throw an error and finish when people try to use X, we end up with just one c-tor. IMO, it is more elegant than to try pick and `#ifdef` individual lines or group of lines.

File: framework/src/timesteppers/TimeStepper.C
Comment (fdkong) [inline framework/src/timesteppers/TimeStepper.C]: Yes,  some applications explicitly use `parameter.xxx<FEProblem *>("_fe_problem")`

**Comment by fdkong** [comment]
> One comment on the EigenSystem and using preprocessor. The following typically worked better for me:
> 
> In .h:
> 
> #if HAVE_A
>   declare class X as if we have A
> #else
>   declare class X as if we do not have A 
>     - keep to c-tor signature the same, but does not require inheritance
>     - does not even need to have the same interface!
> #endif
> 
> In .C
> 
> #if HAVE_A
>   implement class X as if we have A
> #else
>   implement class X as if we do not have A
> #endif
> 
> This obviously depends on how we want X to behave if we don't have A. If our goal is to just throw an error and finish when people try to use X, we end up with just one c-tor. IMO, it is more elegant than to try pick and #ifdef individual lines or group of lines.

I will update according to your suggestions. Thanks a lot!

**Comment by friedmud** [comment]
Ugh... I'm still against this whole `FEProblem`->`FEProblemBase` thing... again... there is a lot of user code with `FEProblem` in it...

**Comment by friedmud** [comment]
Ok - I see now why it's necessary.  I don't see much of a way around it.

Is there any hope that Git is going to do the right thing here to preserve the authorship of all of these thousands of lines of code?

Is there anything we can do to help it out (like add a commit where `FEProblem.C` is renamed to `FEProblemBase.C`)?

**Comment by fdkong** [comment]
> Ugh... I'm still against this whole FEProblem->FEProblemBase thing... again... there is a lot of user code with FEProblem in it...

Could you be more specific? I did not quite get what you said here.  It is fine, if users want to keep using  `FEProblem` (even though it is better to use `FEProblemBase`, in my opinion).

**Comment by fdkong** [comment]
> Ok - I see now why it's necessary. I don't see much of a way around it.
> 
> Is there any hope that Git is going to do the right thing here to preserve the authorship of all of these thousands of lines of code?
> 
> Is there anything we can do to help it out (like add a commit where FEProblem.C is renamed to FEProblemBase.C)?

I have a commit `Rename FEProblem.{C,h} to FEProblemBase.{C, h} 			03f48cd`.  Is it not something you are looking for?

**Comment by friedmud** [comment]
@fdkong It looks like the authorship and paper-trail _are_ intact.  See here: https://github.com/fdkong/moose/blame/nonlinear_eigen_redesign_system_4/framework/src/base/FEProblemBase.C

(then click on "Blame" in the upper right... and let it load)

That was my last objection... so I think this is probably ready to go in.  Any further work can be done as future PRs.

Thanks for hanging in there and putting so much effort into this!

**Comment by YaqiWang** [comment]
I just sent in a PR could interfering with this. Which one should be getting merged first?

**Comment by permcody** [comment]
@YaqiWang - I'd prefer to merge this one first. It's huge, it's been reviewed. It's getting difficult to see small incremental changes at this point. Let's have you rebase your PR after this one is in.

**Comment by permcody** [comment]
image

**Review by andrsd — APPROVED** [review]

**Comment by andrsd** [comment]
Yeah, it is just 10K lines changed in 333 (half-the-devil) files...

**Comment by YaqiWang** [comment]
Congratulations! This will also change my habit of browsing through moose.

**Comment by fdkong** [comment]
Thanks a lot for @andrsd's time on reviewing and merging this tiny change.

Thanks a lot for comments from @friedmud, @permcody and @YaqiWang.

Thanks a lot for @jwpeterson's help on splitting commits.

**Pull request merged**
